### PR TITLE
Port `ExternalTester` from the JS codebase

### PR DESCRIPTION
**Pull request by sachindshinde**
This PR:
- Adds subgraph metadata (known as `FederationMetadata` in the JS codebase) to the schema representation, but doesn't port all the logic there.
- Ports external metadata (known as `ExternalTester` in the JS codebase).

### CODE REVIEW

**Review by goto-bus-stop — CHANGES_REQUESTED** [review]
Implementation looks correct, just some Rusty comments!

File: src/schema/mod.rs [hunk 1]
```diff
@@ -180,42 +183,81 @@ impl FederationSchema {
 
 /// A GraphQL schema with federation data that is known to be valid, and cheap to clone.
 #[derive(Debug, Clone)]
-pub struct ValidFederationSchema(pub(crate) Arc<Valid<FederationSchema>>);
+pub struct ValidFederationSchema {
+    schema: Arc<Valid<FederationSchema>>,
+    subgraph_metadata: Option<SubgraphMetadata>,
```
Comment (goto-bus-stop) [inline src/schema/mod.rs:188]: Oh, this is also an issue, `SubgraphMetadata` is not cheap to clone but is part of `ValidFederationSchema` which (in theory) is.

I think we need to tweak the order of initialisation to fix this and maybe let SubgraphMetadata work on not-necessarily-valid federation schemas.

**Comment by goto-bus-stop** [comment]
I hope it's okay that I addressed my own feedback, that's how I interpreted it in the meeting yesterday, and I was looking for something fun to do!

**Comment by goto-bus-stop** [comment]
`FederationSchema` is used for supergraphs and subgraphs, maybe this should be included in a wrapper type that's only for subgraphs?

**Comment by goto-bus-stop** [comment]
I moved `SubgraphMetadata` to be part of `FederationSchema` instead of `ValidFederationSchema`, but it's only populated in the `ValidFederationSchema` constructor and it can only be accessed through a `ValidFederationSchema`.

I think we should have separate types for subgraph and supergraph schemas, since the code always knows which one it is operating on, and they require different data and different handling.

**Review by lrlna — APPROVED** [review]

**Pull request merged**